core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -661,11 +661,21 @@ class KafkaApis(val requestChannel: RequestChannel,
     val versionId = request.header.apiVersion
     val clientId = request.header.clientId
     val fetchRequest = request.body[FetchRequest]
+    val (topicIds, topicNames) =
+      if (fetchRequest.version() >= 13)
+        metadataCache.topicIdInfo()
+      else
+        (Collections.emptyMap[String, Uuid](), Collections.emptyMap[Uuid, String]())
+
     val fetchContext = fetchManager.newContext(
+      fetchRequest.version,
       fetchRequest.metadata,
-      fetchRequest.fetchData,
-      fetchRequest.toForget,
-      fetchRequest.isFromFollower)
+      fetchRequest.isFromFollower,
+      fetchRequest.fetchDataAndError(topicNames),
+      fetchRequest.toForgetAndIds(topicNames),
```

Comment:
As we resolve all IDs before check permission, does it produce `TOPIC_AUTHORIZATION_FAILED` with topic name even if it has no read permission to such topic (line#717)?

Reply:
I would think so? The goal was for when the IDs are resolved, the path would be the same. And I think we decided that TOPIC_AUTHORIZATION_FAILED is the correct error when we don't have read permissions?

Reply:
Sorry for my unclear comment. My point was that the topic name is included by the response even if there is no read permission to “resolve the topic is”. That looks like a security issue that we expose the topic name.

Reply:
Ah right. I will fix that.

Reply:
I went to fix this but I realized something. Correct me if I'm wrong. The newest version of Fetch will only return topic IDs (not topic names) so we won't expose topic names. The older version of Fetch sends topic names so it will have to send topic names back. I don't think the name is ever incorrectly exposed.